osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs

```diff
@@ -111,50 +115,160 @@ public abstract class DrawableHitObject : SkinReloadableDrawable
+        /// </param>
+        protected DrawableHitObject([CanBeNull] HitObject initialHitObject = null)
         {
-            HitObject = hitObject ?? throw new ArgumentNullException(nameof(hitObject));
+            HitObject = initialHitObject;
         }
 
         [BackgroundDependencyLoader]
         private void load(OsuConfigManager config)
         {
-            userPositionalHitSounds = config.GetBindable<bool>(OsuSetting.PositionalHitSounds);
-            var judgement = HitObject.CreateJudgement();
-
-            Result = CreateResult(judgement);
-            if (Result == null)
-                throw new InvalidOperationException($"{GetType().ReadableName()} must provide a {nameof(JudgementResult)} through {nameof(CreateResult)}.");
+            config.BindWith(OsuSetting.PositionalHitSounds, userPositionalHitSounds);
         }
 
         protected override void LoadAsyncComplete()
         {
             base.LoadAsyncComplete();
 
-            LoadSamples();
+            if (HitObject != null)
+                Apply(HitObject);
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            StartTimeBindable.BindValueChanged(_ => updateState(State.Value, true));
+            comboIndexBindable.BindValueChanged(_ => updateComboColour(), true);
+
+            updateState(ArmedState.Idle, true);
+        }
+
+        /// <summary>
+        /// Applies a new <see cref="HitObject"/> to be represented by this <see cref="DrawableHitObject"/>.
+        /// </summary>
+        /// <param name="hitObject">The <see cref="HitObject"/> to apply.</param>
+        public void Apply(HitObject hitObject)
+        {
+            free();
+
+            HitObject = hitObject ?? throw new InvalidOperationException($"Cannot apply a null {nameof(HitObject)}.");
+
+            // Copy any existing result from the hitobject (required for rewind / judgement revert).
+            Result = HitObject.Result;
+
+            // Ensure this DHO has a result.
+            Result ??= CreateResult(HitObject.CreateJudgement())
+                       ?? throw new InvalidOperationException($"{GetType().ReadableName()} must provide a {nameof(JudgementResult)} through {nameof(CreateResult)}.");
+
+            // Ensure the hitobject has a result.
+            HitObject.Result = Result;
+
+            foreach (var h in HitObject.NestedHitObjects)
+            {
+                var drawableNested = CreateNestedHitObject(h) ?? throw new InvalidOperationException($"{nameof(CreateNestedHitObject)} returned null for {h.GetType().ReadableName()}.");
+
+                drawableNested.OnNewResult += (d, r) => OnNewResult?.Invoke(d, r);
+                drawableNested.OnRevertResult += (d, r) => OnRevertResult?.Invoke(d, r);
+                drawableNested.ApplyCustomUpdateState += (d, j) => ApplyCustomUpdateState?.Invoke(d, j);
```

Comment:
Shouldn't this be

```suggestion
                drawableNested.OnNewResult += onNewResult;
                drawableNested.OnRevertResult += onRevertResult;
                drawableNested.ApplyCustomUpdateState += onApplyCustomUpdateState;
```

It reads nicer for one, but secondly I think that the unsubscribes in `free()` won't have any effect as written as the delegates are not the same (at least that's what I'm seeing in the debugger).

Reply:
Good catch.